### ISSUE TITLE
Possible to export to Android

### ISSUE THREAD

**Issue by ghost** [jiny2001/dcscn-super-resolution#69]
Is it somehow possible to export this model for Android use? I guess it might be doable with Tensorflow Mobile or Tensorflow Lite. Then it would be needed to get hold of checkpoints files (.cpkt) as well as a frozen graph?

**Comment by jiny2001** [comment]
Hi, I'm also curious about porting to mobile. A year ago I have tried TensorFlow RT with freezer graph. But I believe Tensorflow Lite is the we should use? In this case do we just need to convert the checkpoint file??

**Comment by ghost** [comment]
Yes, Tensorflow Lite is the new framework for this. Both checkpoint files (.cpkt) and a graph def (.pb) is needed to get a freezed graph. This freezed graph can be used (and also optimized) to a Tensorflow Lite file.

**Comment by jiny2001** [comment]
So, do you mean you want to add freezed graph output feature on this project?
Or you just want to confirm if it just runs directory under tensorflow lite?

**Comment by ghost** [comment]
Well, I was wondering how you would manage to get a frozen graph from the existing model. The checkpoint files should not be a problem but the graph def (.pb) is not as straight forward.

You can see how the structure for converting looks here (visualized): [https://www.tensorflow.org/lite/convert](https://www.tensorflow.org/lite/convert)

**Comment by jiny2001** [comment]
Well,,, when I tried freeze_graph function on the other project,  I just used this [freeze_graph.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py). And it worked for TF RT. I succeed to make it run on Mac witch C++ in realtime. 
Sorry for not sure about TF Lite.

But I believe you can also check [freeze_graph_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph_test.py) so that you can export it as freezed graph.

**Comment by ghost** [comment]
Okey, thanks! But for example when trying to use the function you provided you need to know output_node_names and in other similar functions also input_nodes. How do you get hold of those? In Tensorboard the graph got a node called output for instance which contains 5 different outputs. Do you know what do provide as output_node_name?

`Y_/output/Mean ?×?×?×1`
`Y_/output/sub ?×?×?×1`
`diff ?×?×?×1`
`gradients/diff_grad/Shape ?×?×?×1`
`Y_/output/dcscn_L8_F96to48_NIN_A64_PS_R1F32`

**Comment by jiny2001** [comment]
Hi, sorry for late reply. I'm so busy these days...
Yes you need the name of the last output node. I think what you want to use is this node.
https://github.com/jiny2001/dcscn-super-resolution/blob/master/DCSCN.py#L265

In this case, the name is just "output". Please try it.

**Comment by iperov** [comment]
I ported to pure keras 
https://github.com/iperov/DeepFaceLab/blob/master/imagelib/DCSCN.py
real model size is 6mb without optimizer weights
works on plaidML backend

**Comment by iperov** [comment]
I applied DCSCN in DeepFaceLab,

before/after gif
![2019-03-28_17-36-57](https://user-images.githubusercontent.com/8076202/55161988-34805480-5180-11e9-83e1-5c5aeda4178a.gif)

**Comment by jiny2001** [comment]
Wow!!
hi iperov, this is AWSOME!! 

I guess you can see the super resolution result better when you use an image which has more clear edges (the sample face image has blurred so it's hard to see the difference). However, the project is also awesome. I pushed a star on your project and will add to my readme that you can use my model via your project. Thank you so much!!

**Comment by iperov** [comment]
@jiny2001
is DCSCN better than for example SRGAN ?
did you compare?
do you know any superresolution network that will work better for blurred images?
do you know any superresolution network that can add details/texture to blurred images?

**Comment by jiny2001** [comment]
Hi,
if you compare DCSCN and SRResNet, SRResNet will have a better result. However, it might be difficult to see the difference from human eyes. And also normal SR network won't work better on blurred image.

SRGAN will normally have less PSNR than other SR networks, however, they generate details/texture and makes it more "natural" so it might work on blurred image. But I haven't tested. 
For the face super resolution, I believe those papers are the best for now. But it only works in certain angle and resolution I guess.
http://openaccess.thecvf.com/content_cvpr_2018/CameraReady/1035.pdf
https://arxiv.org/pdf/1711.10703.pdf

Anyway, I guess you can try SRGAN since there are some implementations in github. Thx.

**Comment by iperov** [comment]
my opinion is that all SRs are actually trained to restorate downscaled images, so these images have high density of "details per pixel". 
But real world images are not downscaled images :D 

especially bicubic downsampled image has more information about upscaled image than bilinear or nearest-neighbor. So why to train SR to restorate images that already have more information than real world image?
My suggestion to train x2 DCSCN upscale network on x4 nearest neighbor downscaled (random offset 1-4px), then x2 bilinear upscaled images

**Comment by jiny2001** [comment]
You are right. I agree I should train from bilinear. I'm just using bicubic because every other people use it and I can compare the performance with it. 
let me know if you can point me a nice training data set. I can try it with bilinear downsampling. ;)

**Comment by ricardobnjunior** [comment]
> Okey, thanks! But for example when trying to use the function you provided you need to know output_node_names and in other similar functions also input_nodes. How do you get hold of those? In Tensorboard the graph got a node called output for instance which contains 5 different outputs. Do you know what do provide as output_node_name?
> 
> `Y_/output/Mean ?×?×?×1`
> `Y_/output/sub ?×?×?×1`
> `diff ?×?×?×1`
> `gradients/diff_grad/Shape ?×?×?×1`
> `Y_/output/dcscn_L8_F96to48_NIN_A64_PS_R1F32`

have you found the input node name?

**Comment by iperov** [comment]
![2019-04-14_21-18-47](https://user-images.githubusercontent.com/8076202/56096565-f6fb2580-5efa-11e9-82b6-ef630ed0f3b6.gif)

**Comment by jiny2001** [comment]
Hi @ricardobnjunior, sorry for late reply. 
The input node is the one which is defined as placeholder, the name is "x". I guess you have found this already though...

Hi @iperov, the result is really awesome!! Great!!
I guess much much smaller model will also work if it focused on face super resolution. Which will greatly improve processing speed. Will try it when i have a time. Thx!!

**Comment by ricardobnjunior** [comment]
> Hi @ricardobnjunior, sorry for late reply.
> The input node is the one which is defined as placeholder, the name is "x". I guess you have found this already though...
> 
> Hi @iperov, the result is really awesome!! Great!!
> I guess much much smaller model will also work if it focused on face super resolution. Which will greatly improve processing speed. Will try it when i have a time. Thx!!

Yes, I found!
Thank you for reply ;)

**Comment by djalusic** [comment]
@ricardobnjunior, have you succeeded in converting the model to TFLite?

**Comment by nkache** [comment]
> my opinion is that all SRs are actually trained to restorate downscaled images, so these images have high density of "details per pixel".
> But real world images are not downscaled images :D
> 
> especially bicubic downsampled image has more information about upscaled image than bilinear or nearest-neighbor. So why to train SR to restorate images that already have more information than real world image?
> My suggestion to train x2 DCSCN upscale network on x4 nearest neighbor downscaled (random offset 1-4px), then x2 bilinear upscaled images

@iperov Could you please explain further about the rationale behind suggesting this -  "x4 nearest neighbor downscaled (random offset 1-4px), then x2 bilinear upscaled images"

**Comment by iperov** [comment]
@Nik-Khandelwal 

main idea is remove more information from downscaled image about it upscaled version, therefore neural network will not learn to fetch that info
